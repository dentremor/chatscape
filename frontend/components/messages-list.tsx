import {
  List,
  ListItem,
  Card,
  CardBody,
  EmptyState,
  EmptyStateVariant,
  EmptyStateBody,
  Skeleton,
} from "@patternfly/react-core";
import { RefObject, Fragment } from "react";

export const MessagesList = ({
  messages,
  lastMessageRef,
}: {
  messages?: {
    them?: boolean;
    body: string;
    date: Date;
  }[];
  lastMessageRef: RefObject<HTMLSpanElement> | undefined;
}) => (
  <List isPlain>
    {messages ? (
      messages.length > 0 ? (
        messages.map((message, i) => (
          <Fragment key={i}>
            <ListItem>
              <Card
                isCompact
                isRounded
                className={
                  "pf-c-card--message " +
                  (message.them ? "pf-c-card--them" : "pf-c-card--us")
                }
              >
                <CardBody>{message.body}</CardBody>
              </Card>
            </ListItem>

            {(messages[i + 1] &&
              Math.abs(
                message.date.getTime() - messages[i + 1].date.getTime()
              ) /
                (1000 * 60 * 60) >
                2) ||
            i === messages.length - 1 ? (
              <ListItem>
                <span
                  ref={i === messages.length - 1 ? lastMessageRef : undefined}
                  className="pf-c-date"
                >
                  {`Today ${message.date
                    .getHours()
                    .toString()
                    .padStart(2, "0")}:${message.date
                    .getMinutes()
                    .toString()
                    .padStart(2, "0")}`}
                </span>
              </ListItem>
            ) : null}
          </Fragment>
        ))
      ) : (
        <ListItem>
          <EmptyState variant={EmptyStateVariant.xs}>
            <EmptyStateBody className="pf-u-mt-0">
              No messages yet
            </EmptyStateBody>
          </EmptyState>
        </ListItem>
      )
    ) : (
      <>
        <ListItem>
          <Skeleton screenreaderText="Loading messages" width="90%" />
        </ListItem>

        <ListItem>
          <Skeleton screenreaderText="Loading messages" width="66%" />
        </ListItem>

        <ListItem>
          <Skeleton screenreaderText="Loading messages" width="77%" />
        </ListItem>

        <ListItem>
          <Skeleton screenreaderText="Loading messages" width="33%" />
        </ListItem>

        <ListItem className="pf-u-mt-3xl">
          <Skeleton
            screenreaderText="Loading messages"
            width="66%"
            className="pf-u-ml-auto"
          />
        </ListItem>

        <ListItem>
          <Skeleton
            screenreaderText="Loading messages"
            width="33%"
            className="pf-u-ml-auto"
          />
        </ListItem>
      </>
    )}
  </List>
);
