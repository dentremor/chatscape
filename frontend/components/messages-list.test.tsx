import { render } from "@testing-library/react";
import { createRef } from "react";
import { MessagesList } from "./messages-list";

describe("MessagesList", () => {
  it("renders with messages set", () => {
    const messages = [
      {
        them: true,
        body: "Hello there!",
        date: new Date(2023, 6, 13, 8, 30),
      },
      {
        them: false,
        body: "Hi, how can I help you?",
        date: new Date(2023, 6, 13, 8, 35),
      },
    ];
    expect(
      render(<MessagesList messages={messages} lastMessageRef={createRef()} />)
    ).toMatchSnapshot();
  });

  it("renders without messages set", () => {
    expect(
      render(<MessagesList lastMessageRef={createRef()} />)
    ).toMatchSnapshot();
  });
});
