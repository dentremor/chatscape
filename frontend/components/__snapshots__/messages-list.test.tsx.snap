// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessagesList renders with messages set 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <ul
        class="pf-c-list pf-m-plain"
      >
        <li
          class=""
        >
          <article
            class="pf-c-card pf-m-compact pf-m-rounded pf-c-card--message pf-c-card--them"
            data-ouia-component-id="OUIA-Generated-Card-1"
            data-ouia-component-type="PF4/Card"
            data-ouia-safe="true"
            id=""
          >
            <div
              class="pf-c-card__body"
            >
              Hello there!
            </div>
          </article>
        </li>
        <li
          class=""
        >
          <article
            class="pf-c-card pf-m-compact pf-m-rounded pf-c-card--message pf-c-card--us"
            data-ouia-component-id="OUIA-Generated-Card-2"
            data-ouia-component-type="PF4/Card"
            data-ouia-safe="true"
            id=""
          >
            <div
              class="pf-c-card__body"
            >
              Hi, how can I help you?
            </div>
          </article>
        </li>
        <li
          class=""
        >
          <span
            class="pf-c-date"
          >
            Today 08:35
          </span>
        </li>
      </ul>
    </div>
  </body>,
  "container": <div>
    <ul
      class="pf-c-list pf-m-plain"
    >
      <li
        class=""
      >
        <article
          class="pf-c-card pf-m-compact pf-m-rounded pf-c-card--message pf-c-card--them"
          data-ouia-component-id="OUIA-Generated-Card-1"
          data-ouia-component-type="PF4/Card"
          data-ouia-safe="true"
          id=""
        >
          <div
            class="pf-c-card__body"
          >
            Hello there!
          </div>
        </article>
      </li>
      <li
        class=""
      >
        <article
          class="pf-c-card pf-m-compact pf-m-rounded pf-c-card--message pf-c-card--us"
          data-ouia-component-id="OUIA-Generated-Card-2"
          data-ouia-component-type="PF4/Card"
          data-ouia-safe="true"
          id=""
        >
          <div
            class="pf-c-card__body"
          >
            Hi, how can I help you?
          </div>
        </article>
      </li>
      <li
        class=""
      >
        <span
          class="pf-c-date"
        >
          Today 08:35
        </span>
      </li>
    </ul>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`MessagesList renders without messages set 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <ul
        class="pf-c-list pf-m-plain"
      >
        <li
          class=""
        >
          <div
            class="pf-c-skeleton"
            style="--pf-c-skeleton--Width: 90%;"
          >
            <span
              class="pf-u-screen-reader"
            >
              Loading messages
            </span>
          </div>
        </li>
        <li
          class=""
        >
          <div
            class="pf-c-skeleton"
            style="--pf-c-skeleton--Width: 66%;"
          >
            <span
              class="pf-u-screen-reader"
            >
              Loading messages
            </span>
          </div>
        </li>
        <li
          class=""
        >
          <div
            class="pf-c-skeleton"
            style="--pf-c-skeleton--Width: 77%;"
          >
            <span
              class="pf-u-screen-reader"
            >
              Loading messages
            </span>
          </div>
        </li>
        <li
          class=""
        >
          <div
            class="pf-c-skeleton"
            style="--pf-c-skeleton--Width: 33%;"
          >
            <span
              class="pf-u-screen-reader"
            >
              Loading messages
            </span>
          </div>
        </li>
        <li
          class="pf-u-mt-3xl"
        >
          <div
            class="pf-c-skeleton pf-u-ml-auto"
            style="--pf-c-skeleton--Width: 66%;"
          >
            <span
              class="pf-u-screen-reader"
            >
              Loading messages
            </span>
          </div>
        </li>
        <li
          class=""
        >
          <div
            class="pf-c-skeleton pf-u-ml-auto"
            style="--pf-c-skeleton--Width: 33%;"
          >
            <span
              class="pf-u-screen-reader"
            >
              Loading messages
            </span>
          </div>
        </li>
      </ul>
    </div>
  </body>,
  "container": <div>
    <ul
      class="pf-c-list pf-m-plain"
    >
      <li
        class=""
      >
        <div
          class="pf-c-skeleton"
          style="--pf-c-skeleton--Width: 90%;"
        >
          <span
            class="pf-u-screen-reader"
          >
            Loading messages
          </span>
        </div>
      </li>
      <li
        class=""
      >
        <div
          class="pf-c-skeleton"
          style="--pf-c-skeleton--Width: 66%;"
        >
          <span
            class="pf-u-screen-reader"
          >
            Loading messages
          </span>
        </div>
      </li>
      <li
        class=""
      >
        <div
          class="pf-c-skeleton"
          style="--pf-c-skeleton--Width: 77%;"
        >
          <span
            class="pf-u-screen-reader"
          >
            Loading messages
          </span>
        </div>
      </li>
      <li
        class=""
      >
        <div
          class="pf-c-skeleton"
          style="--pf-c-skeleton--Width: 33%;"
        >
          <span
            class="pf-u-screen-reader"
          >
            Loading messages
          </span>
        </div>
      </li>
      <li
        class="pf-u-mt-3xl"
      >
        <div
          class="pf-c-skeleton pf-u-ml-auto"
          style="--pf-c-skeleton--Width: 66%;"
        >
          <span
            class="pf-u-screen-reader"
          >
            Loading messages
          </span>
        </div>
      </li>
      <li
        class=""
      >
        <div
          class="pf-c-skeleton pf-u-ml-auto"
          style="--pf-c-skeleton--Width: 33%;"
        >
          <span
            class="pf-u-screen-reader"
          >
            Loading messages
          </span>
        </div>
      </li>
    </ul>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
